native/core/src/execution/datafusion/expressions/cast.rs

```diff
@@ -501,18 +504,24 @@ impl Cast {
         let array = array_with_timezone(array, self.timezone.clone(), Some(to_type))?;
         let from_type = array.data_type().clone();
 
-        // unpack dictionary string arrays first
-        // TODO: we are unpacking a dictionary-encoded array and then performing
-        // the cast. We could potentially improve performance here by casting the
-        // dictionary values directly without unpacking the array first, although this
-        // would add more complexity to the code
         let array = match &from_type {
             DataType::Dictionary(key_type, value_type)
                 if key_type.as_ref() == &DataType::Int32
                     && (value_type.as_ref() == &DataType::Utf8
                         || value_type.as_ref() == &DataType::LargeUtf8) =>
             {
-                cast_with_options(&array, value_type.as_ref(), &CAST_OPTIONS)?
+                let dict_array = array
+                    .as_any()
+                    .downcast_ref::<DictionaryArray<Int32Type>>()
+                    .expect("Expected a dictionary array");
+                let values = dict_array.values();
+                let cast_values = self.cast_array(values.clone())?;
+                let cast_array = Arc::new(DictionaryArray::<Int32Type>::new(
+                    dict_array.keys().clone(),
+                    cast_values,
+                )) as ArrayRef;
```

Comment:
This code looks good and is converting the input dictionary array to the output dictionary array that we want